### PR TITLE
Replace BIO_printf with ASN1_STRING_print in GENERAL_NAME_print

### LINKED ISSUE #8997: GENERAL_NAME_print mishandled embedded nuls in IA5 string values
See [https://github.com/openssl/openssl/blob/master/crypto/x509v3/v3_alt.c](url)

```
    case GEN_EMAIL:
        BIO_printf(out, "email:%s", gen->d.ia5->data);
        break;

    case GEN_DNS:
        BIO_printf(out, "DNS:%s", gen->d.ia5->data);
        break;

    case GEN_URI:
        BIO_printf(out, "URI:%s", gen->d.ia5->data);
        break;
```

Each of these instances should be using ASN1_STRING_print to print the value rather than printing the raw ia5 string data. If there is an embedded end-of-string (\x00) character the user will be unaware.

For code using GENERAL_NAME_print to handle the various formats and then post-processing for checking logic, the wrong handling will occur.

For users working with the command line applications, incorrect output will occur.

If we have embedded end-of-string values in the subject or issuer name, those are displayed correctly. For general ASN1_STRING handling we fold special characters to a period (.) which is another way of indicating there is more data. 

Reproduce the issue with a certificate created with a SAN with DNS and an embedded \x00 within the DNS name (yes this is an invalid usage - however we shouldn't silently truncate the output which may lead to unexpected handling).

### PR DESCRIPTION
**Pull request by acheevbhagat**
Fixes #8997

### CODE REVIEW

File: crypto/x509/v3_alt.c [hunk 1]
```diff
@@ -157,15 +157,18 @@ int GENERAL_NAME_print(BIO *out, GENERAL_NAME *gen)
         break;
 
     case GEN_EMAIL:
-        BIO_printf(out, "email:%s", gen->d.ia5->data);
+        BIO_printf(out, "email:");
+	ASN1_STRING_print(out, gen->d.ia5);
```
Comment (mspncp) [inline crypto/x509/v3_alt.c]: whitespace error: replace tab by four spaces

File: crypto/x509/v3_alt.c [hunk 2]
```diff
@@ -157,15 +157,18 @@ int GENERAL_NAME_print(BIO *out, GENERAL_NAME *gen)
...
         break;
 
     case GEN_DNS:
-        BIO_printf(out, "DNS:%s", gen->d.ia5->data);
+        BIO_printf(out, "DNS:");
+	ASN1_STRING_print(out, gen->d.ia5);
```
Comment (mspncp) [inline crypto/x509/v3_alt.c]: whitespace error: replace tab by four spaces

File: crypto/x509/v3_alt.c [hunk 3]
```diff
@@ -157,15 +157,18 @@ int GENERAL_NAME_print(BIO *out, GENERAL_NAME *gen)
...
         break;
 
     case GEN_URI:
-        BIO_printf(out, "URI:%s", gen->d.ia5->data);
+        BIO_printf(out, "URI:");
+	ASN1_STRING_print(out, gen->d.ia5);
```
Comment (mspncp) [inline crypto/x509/v3_alt.c]: whitespace error: replace tab by four spaces

File: crypto/x509/v3_alt.c (hunk 1 above)
Comment (mspncp) [reply crypto/x509/v3_alt.c]: correction: replace tab bei eight spaces  (three locations)

**Comment by t8m** [comment]
@acheevbhagat To give a hint how to do the fixups - just use `git commit --fixup HEAD` to commit the fixes to the PR.

**Comment by mspncp** [comment]
Merged. Thank you @acheevbhagat for your contribution!

bab6046146 Replace BIO_printf with ASN1_STRING_print in GENERAL_NAME_print [master]
7febec9485 Replace BIO_printf with ASN1_STRING_print in GENERAL_NAME_print [1.1.1]

**Comment by t-j-h** [comment]
@mspncp  I think this should also go into 1.0.2 - it is a security related issue - and it applies there as well.

**Comment by mspncp** [comment]
> @mspncp I think this should also go into 1.0.2 - it is a security related issue - and it applies there as well.

Ok, I can do it.

I already verified that commit bab6046146c4fc8f088c1aaca11598ede7609c04 cherry-picks to 1.0.2 without conflicts. But before pushing it, I'd like to wait 24 hours to give other @openssl/omc members the opportunity to express their opinions about whether the change qualifies as security fix or not.

**Comment by mattcaswell** [comment]
Why is this a security fix?

**Comment by t-j-h** [comment]
Any application using the print function to compare names in certificates will match incorrectly if there is an embedded 0x00 in the ia5string. And human looking at the values will see the wrong information.

**Comment by mattcaswell** [comment]
Ok then.

**Comment by mspncp** [comment]
Given two OMC approvals for the cherry-pick, I'll reduce the waiting time to 6 hours, i.e., I'll do it this evening.

**Comment by mspncp** [comment]
And finally, merged to 1.0.2

8479e9e973 Replace BIO_printf with ASN1_STRING_print in GENERAL_NAME_print [1.0.2]

**Pull request closed without merging**